main.go

```diff
@@ -235,39 +353,177 @@ func main() {
+	// encryption key.
+	newCacheFunc := cache.BuilderWithOptions(
+		cache.Options{
+			SelectorsByObject: cache.SelectorsByObject{
+				&v1.Secret{}: {
+					Field: fields.SelectorFromSet(fields.Set{"metadata.name": secretsync.SecretName}),
+				},
+			},
+		},
+	)
+
+	// Set the manager options
+	options.HealthProbeBindAddress = healthAddr
+	options.LeaderElectionID = "policy-syncer2.open-cluster-management.io"
+	options.LeaderElectionConfig = managedCfg
+	options.NewCache = newCacheFunc
+
+	// Create a new manager to provide shared dependencies and start components
+	mgr, err := ctrl.NewManager(hubCfg, options)
+	if err != nil {
+		log.Error(err, "Failed to start manager")
+		os.Exit(1)
+	}
+
+	// Setup all Controllers
+	if err = (&specsync.PolicyReconciler{
+		HubClient:       mgr.GetClient(),
+		ManagedClient:   managedClient,
+		ManagedRecorder: managedRecorder,
+		Scheme:          mgr.GetScheme(),
+		TargetNamespace: tool.Options.ClusterNamespace,
+	}).SetupWithManager(mgr); err != nil {
+		log.Error(err, "Unable to create the controller", "controller", specsync.ControllerName)
+		os.Exit(1)
+	}
+
+	if err = (&secretsync.SecretReconciler{
+		Client:          mgr.GetClient(),
+		ManagedClient:   managedClient,
+		Scheme:          mgr.GetScheme(),
+		TargetNamespace: tool.Options.ClusterNamespace,
+	}).SetupWithManager(mgr); err != nil {
+		log.Error(err, "Unable to create the controller", "controller", secretsync.ControllerName)
+		os.Exit(1)
+	}
+
+	// use config check
+	configChecker, err := addonutils.NewConfigChecker("policy-syncer2", tool.Options.HubConfigFilePathName)
+	if err != nil {
+		log.Error(err, "unable to setup a configChecker")
+		os.Exit(1)
+	}
+
+	//+kubebuilder:scaffold:builder
+	if err := mgr.AddHealthzCheck("healthz", configChecker.Check); err != nil {
+		log.Error(err, "unable to set up health check")
+		os.Exit(1)
+	}
+
+	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+		log.Error(err, "unable to set up ready check")
+		os.Exit(1)
+	}
+
+	return mgr
+}
+
+// startHealthProxy responds to /healthz and /readyz HTTP requests and combines the status together of the input
+// addresses representing the managers.
+func startHealthProxy(ctx context.Context, wg *sync.WaitGroup, addresses ...string) {
+	for _, endpoint := range []string{"/healthz", "/readyz"} {
+		http.HandleFunc(endpoint, func(w http.ResponseWriter, r *http.Request) {
+			for _, address := range addresses {
+				req, err := http.NewRequestWithContext(
+					ctx, http.MethodGet, fmt.Sprintf("http://%s%s", address, endpoint), nil,
+				)
+				if err != nil {
+					http.Error(w, fmt.Sprintf("manager: %s", err.Error()), http.StatusInternalServerError)
+
+					return
+				}
+
+				resp, err := http.DefaultClient.Do(req)
+				if err != nil {
+					http.Error(w, fmt.Sprintf("manager: %s", err.Error()), http.StatusInternalServerError)
+
+					return
+				}
+
+				defer func() { resp.Body.Close() }()
```

Comment:
Nit: Based on an example in https://pkg.go.dev/net/http#Server, this could be simplified
```suggestion
				defer resp.Body.Close()
```